text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,114 @@
+# Handling of Equivocation Catastrophes
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0000](https://github.com/casperlabs/ceps/pull/0000)
+
+This CEP describes a proposal for the process of handling equivocation catastrophes in Highway consensus (or, more generally: handling situations that break the assumptions behind the security of the consensus algorithm being used).
+
+## Motivation
+
+[motivation]: #motivation
+
+Every consensus algorithm's safety guarantees rely on some assumptions being upheld. In the case of Highway, the main such assumption is about the validators not equivocating, that is: not producing conflicting votes. The algorithm can withstand some amount of equivocations, but if the total weight of equivocating validators exceeds a threshold called the Fault Tolerance Threshold (FTT), they could revert finalized blocks, which could lead for example to double spending. We call such a situation an "equivocation catastrophe" and it is something that cannot be handled automatically (as it would effectively contradict well-established results about the limitations of consensus algorithms in general). Thus, we require some kind of an external mechanism for dealing with such situations, should they ever happen.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The procedure for handling the equivocation catastrophes will have to consist of two main parts: what the software is expected to do in such a case, and what are the manual actions required to proceed.
+
+The obvious first thing that should happen in the case of an equivocation catastrophe is that the software should alert the users about the situation. The node software is aware of the configured FTT, so when it detects equivocations, it can notice if the total weight of the equivocators exceeds the current FTT. When it is exceeded, the users should be made aware of it.
+
+In order to make sure that as many validators as possible become aware of all the equivocations eventually, the operation of the nodes shouldn't stop. Finalizing blocks should be halted, as finalization ceases to mean anything when it can be reverted, but the nodes should continue gossiping votes.
```

Comment:
How will the validators know when all of them have reached the same protocol state? One where all equivocators are known all everyone has seen all of the consensus messages?

Reply:
Good question. I wonder if a good solution here would be to stop creating new votes altogether, but keep gossiping the ones we have until the network looks saturated from our perspective?